### PR TITLE
Reduce default values of the quest file redux

### PR DESCRIPTION
**Pull request by WaitingIdly**
changes in this PR:
- add default values for most tasks and rewards.
- when saving to the database, only save non-default values. this reduces the size of the questbook significantly - reportedly around 50% by both Nomifactory and DJ2.
- deprecate `NBTReplaceUtil` and add `NBTUtil`, which simplifies code related to these operations.
- fix an error where trying to load quests from the new database would error due to simply not clearing current quests (this used to happen when calling `QuestDatabase.INSTANCE.readFromNBT(nbt1.getTagList("questDatabase", 10), false);`, with the `false` causes a `QuestDatabase.INSTANCE.reset();` to be called. however, due to breaking the database into separate files, this method is no longer called, and reset didnt get copied over).
- similarly, readded `questID`, `lineID`, and `order` tags to the quest database and questline when running `save`, as those are called via the `writeToNBT` methods, which are unused in this context (only used in `saveLegacy`).
- update the questbook format from `2.0.0` to `2.1.0`.

difference from #109:
- expanded significantly on what `NBTUtil` does.
- added more default things, fixed a few situations where NBT data tags were changed.
- apply reduction to saving and loading in the new save/load layout.
- fix some whitespace issues.

supersedes #109

### CODE REVIEW

File: src/main/java/betterquesting/NBTUtil.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        return baseTag; // Either isn't a string or doesn't contain one
+    }
+
+    public static void setTag(NBTTagCompound tag, String key, NBTBase value, boolean reduce) {
+        if (reduce && value.isEmpty()) return;
+        tag.setTag(key, value);
+    }
+
+    public static boolean getBoolean(NBTTagCompound tag, String key, boolean defaultValue) {
+        return tag.hasKey(key, 99) ? tag.getBoolean(key) : defaultValue;
```
Comment (TechLord22) [inline src/main/java/betterquesting/NBTUtil.java]: The 99s in this file can be `Constants.NBT.TAG_ANY_NUMERIC` from `net.minecraftforge.common.util.Constants`.

File: src/main/java/betterquesting/NBTUtil.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        return tag.hasKey(key, 99) ? tag.getFloat(key) : defaultValue;
+    }
+
+    public static void setFloat(NBTTagCompound tag, String key, float value, float defaultValue, boolean reduce) {
+        if (reduce && value == defaultValue) return;
+        tag.setFloat(key, value);
+    }
+
+    public static String getString(NBTTagCompound tag, String key, String defaultValue) {
+        return tag.hasKey(key, 8) ? tag.getString(key) : defaultValue;
```
Comment (TechLord22) [inline src/main/java/betterquesting/NBTUtil.java]: The 8s in this file can use `Constants.NBT.TAG_STRING`

File: src/main/java/betterquesting/api/properties/basic/PropertyTypeItemStack.java [hunk 3]
```diff
@@ -25,11 +27,23 @@ public NBTBase writeValue(BigItemStack value) {
...
+            value.writeToNBT(nbt, false);
         }
 
         return nbt;
     }
+
+    @Override
+    public NBTBase reduceNBT(NBTBase nbt) {
+        BigItemStack value;
+        if (nbt == null || nbt.getId() != 10) {
```
Comment (TechLord22) [inline src/main/java/betterquesting/api/properties/basic/PropertyTypeItemStack.java]: Can do `nbt.getId() != Constants.NBT.TAG_COMPOUND` instead.

File: src/main/java/betterquesting/commands/admin/QuestCommandDefaults.java [hunk 4]
```diff
@@ -191,7 +188,9 @@ public static void save(@Nullable ICommandSender sender, String databaseName, Fi
             String questLineNameTranslated = QuestTranslation.translate(questLineName);
 
             File questLineFile = new File(questLineDir, buildFileName.apply(questLineNameTranslated, questLineId) + ".json");
-            NBTTagCompound questLineTag = questLine.writeToNBT(new NBTTagCompound(), null);
+            NBTTagCompound questLineTag = questLine.writeToNBT(new NBTTagCompound(), null, true);
+            questLineTag.setInteger("lineID", questLineId);
+            questLineTag.setInteger("order", QuestLineDatabase.INSTANCE.getOrderIndex(entry.getID()));
```
Comment (TechLord22) [inline src/main/java/betterquesting/commands/admin/QuestCommandDefaults.java:194]: Why are these needed now?

File: src/main/java/betterquesting/commands/admin/QuestCommandDefaults.java [hunk 5]
```diff
@@ -230,7 +229,8 @@ public static void save(@Nullable ICommandSender sender, String databaseName, Fi
                 return;
             }
 
-            NBTTagCompound questTag = quest.writeToNBT(new NBTTagCompound());
+            NBTTagCompound questTag = quest.writeToNBT(new NBTTagCompound(), true);
+            questTag.setInteger("questID", questId);
```
Comment (TechLord22) [inline src/main/java/betterquesting/commands/admin/QuestCommandDefaults.java:234]: Why is this needed now?

File: src/main/java/betterquesting/commands/admin/QuestCommandDefaults.java [hunk 6]
```diff
@@ -311,14 +311,19 @@ public static void load(@Nullable ICommandSender sender, @Nullable String databa
                 .forEach(questLineDatabase::appendTag);
 
         QuestLineDatabase.INSTANCE.readFromNBT(questLineDatabase, false);
+        QuestDatabase.INSTANCE.reset();
```
Comment (TechLord22) [inline src/main/java/betterquesting/commands/admin/QuestCommandDefaults.java:315]: Why is this needed now?

File: src/main/java/betterquesting/commands/admin/QuestCommandDefaults.java [hunk 7]
```diff
@@ -311,14 +311,19 @@ public static void load(@Nullable ICommandSender sender, @Nullable String databa
...
 
         File questDir = new File(dataDir, QUEST_DIR);
         try (Stream<Path> paths = Files.walk(questDir.toPath())) {
             paths.filter(Files::isRegularFile).forEach(
                     path -> {
                         File questFile = path.toFile();
                         NBTTagCompound questTag = readNbt.apply(questFile);
-                        int questId = Integer.parseInt(questFile.getName().replaceAll("[^0-9]+", ""));
+                        int questId = questTag.hasKey("questID", 99) ? questTag.getInteger("questID") : -1;
```
Comment (TechLord22) [inline src/main/java/betterquesting/commands/admin/QuestCommandDefaults.java]: Can use `Constants`.

File: src/main/java/betterquesting/commands/admin/QuestCommandDefaults.java (hunk 6 above)
Comment (WaitingIdly) [reply src/main/java/betterquesting/commands/admin/QuestCommandDefaults.java:315]: this is a result of an incorrect migration of code from the old format the results in errors.
in the old format, it calls `QuestDatabase.INSTANCE.readFromNBT(nbt1.getTagList("questDatabase", 10), false)`.
the boolean there is `merge`, and inside `readFromNBT` reset is called if it is false `if (!merge) this.reset();`.
thus, reset should be called when loading. however, when adding the modern format, this was missed, and it does not call `QuestDatabase.INSTANCE.reset()`.
this results in an error whenever you try to load quests into a world that already has quests that match an already created quest id.

File: src/main/java/betterquesting/commands/admin/QuestCommandDefaults.java (hunk 4 above)
Comment (WaitingIdly) [reply src/main/java/betterquesting/commands/admin/QuestCommandDefaults.java:194]: this is a result of an incorrect migration of code from the old format the results in data not being saved to the file.
in the old format, it calls `QuestLineDatabase.INSTANCE.writeToNBT(new NBTTagList(), null, true)`, and inside that method:
```java
            jObj.setInteger("lineID", entry.getID());
            jObj.setInteger("order", getOrderIndex(entry.getID()));
```
the `lineID` and `order` are saved. however, this was not copied to the new save format, and so saving currently does not save the line order - which results in situations such as this (from Dimension Hopper)
image

File: src/main/java/betterquesting/commands/admin/QuestCommandDefaults.java (hunk 5 above)
Comment (WaitingIdly) [reply src/main/java/betterquesting/commands/admin/QuestCommandDefaults.java:234]: this has the same reason as the `lineID` and `order` properties, but its merely a missing value from the file that ought to be there instead of causing a bug.

**Review by TechLord22 — APPROVED** [review]
